File: typescript/bob/bob.ts
const isUppercase = (letter: string): boolean => /[A-Z]/.test(letter);
const isLetter = (letter: string): boolean => /[a-z]/i.test(letter);

export function hey(message: string): string {
  const tokenizedMessage = message.trim().split('')

  const hasLetters = tokenizedMessage.some(isLetter);
  const allUppercase = tokenizedMessage.filter(isLetter).every(isUppercase);
  const questionLike = tokenizedMessage[tokenizedMessage.length - 1] === '?';

  if (tokenizedMessage.length === 0) {
    return 'Fine. Be that way!';
  }

  if (allUppercase && !questionLike && hasLetters) {
    return 'Whoa, chill out!';
  }

  if (allUppercase && questionLike && hasLetters) {
    return "Calm down, I know what I'm doing!";
  }

  if (questionLike) {
    return 'Sure.';
  }

  return 'Whatever.';
}

File: erlang/bob/src/bob.erl
-module(bob).

-export([response/1]).


-spec is_letter(char()) -> boolean().
is_letter(Character) when Character >= $A andalso Character =< $Z -> true;
is_letter(Character) when Character >= $a andalso Character =< $z -> true;
is_letter(_Character)                                             -> false.

-spec is_uppercase(char()) -> boolean().
is_uppercase(Character) when Character >= $A andalso Character =< $Z -> true;
is_uppercase(_Character)                                             -> false.

-spec is_question(string()) -> boolean().
is_question([])        -> false;
is_question(Statement) -> lists:last(Statement) =:= $?.

-spec response(string()) -> string().
response(Statement) ->
    TrimmedStatement = string:trim(Statement),

    StatementLength = length(TrimmedStatement),

    SanitizedStatement = lists:filter(fun is_letter/1, TrimmedStatement),
    HasLetters = length(SanitizedStatement) > 0,

    AllUppercase = lists:all(fun is_uppercase/1, SanitizedStatement),

    QuestionLike = is_question(TrimmedStatement),

    if
        StatementLength =:= 0 -> "Fine. Be that way!";

        HasLetters andalso AllUppercase andalso QuestionLike -> "Calm down, I know what I'm doing!";
        HasLetters andalso AllUppercase andalso not QuestionLike -> "Whoa, chill out!";
        QuestionLike -> "Sure.";

        true -> "Whatever."
    end.
